### PR TITLE
CLOUDP-269300: Bump to Go 1.23

### PR DESCRIPTION
**Pull request by roothorp**
Including the relevant golangci version bump to support Go 1.23.
Also includes other miscellaneous version updates for devbox (these will be automated similar to dependabot in future).

wrt `exportloopref`:
> Since Go1.22 (loopvar) this linter is no longer relevant. Replaced by copyloopvar.

### All Submissions:

* [ ] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

**Comment by josvazg** [comment]
@roothorp I would make this a draft until we figure out the licenses issue. BTW, I can take this PR if you want.

**Comment by josvazg** [comment]
Taking this from @roothorp will make it a draft while I figure out the licenses issue

File: Makefile [hunk 1]
```diff
@@ -126,9 +127,9 @@ all: manager ## Build all binaries
 build-licenses.csv: go.mod ## Track licenses in a CSV file
 	@echo "Tracking licenses into file $@"
 	@echo "========================================"
-	GOOS=linux GOARCH=amd64 go mod download
-	# https://github.com/google/go-licenses/issues/244
-	GOTOOLCHAIN=local GOOS=linux GOARCH=amd64 $(GO_LICENSES) csv --include_tests $(BASE_GO_PACKAGE)/... > licenses.csv
+	export GOOS=linux
+	export GOARCH=amd64
+	go run github.com/google/$(GO_LICENSES)@v$(GO_LICENSES_VERSION) csv --include_tests $(BASE_GO_PACKAGE)/... > licenses.csv
```
Comment (josvazg) [inline Makefile:132]: Not the prettiest fix, but the only one I could find. 

From here:
https://github.com/google/go-licenses/issues/149#issuecomment-2208882223

Seems we won't be able to use devbox/nix for the go-licenses binary and instead need to use it as a module dep. 

cc @roothorp

**Review by josvazg — APPROVED** [review]

File: pkg/controller/atlasproject/custom_roles.go [hunk 2]
```diff
@@ -210,7 +210,7 @@ func syncCustomRolesStatus(ctx *workflow.Context, desiredCustomRoles []customrol
 			statuses = append(statuses, stat)
 
 			if stat.Status == status.CustomRoleStatusFailed {
-				err = errors.Join(err, fmt.Errorf(stat.Error))
+				err = errors.Join(err, fmt.Errorf("%s", stat.Error))
```
Comment (s-urbaniak) [inline pkg/controller/atlasproject/custom_roles.go:213]:
```suggestion
				err = errors.Join(err, fmt.Errorf("%w", stat.Error))
```

File: pkg/controller/atlasproject/custom_roles.go [hunk 3]
```diff
@@ -220,7 +220,7 @@ func syncCustomRolesStatus(ctx *workflow.Context, desiredCustomRoles []customrol
...
```
Comment (s-urbaniak) [inline pkg/controller/atlasproject/custom_roles.go:223]:
```suggestion
				err = errors.Join(err, fmt.Errorf("%w", stat.Error))
```

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
one nit, else lgtm 👍

File: pkg/controller/atlasproject/custom_roles.go [hunk 4]
```diff
@@ -200,7 +200,7 @@ func syncCustomRolesStatus(ctx *workflow.Context, desiredCustomRoles []customrol
 		if ok {
 			if stat.Status == status.CustomRoleStatusFailed {
 				statuses = append(statuses, stat)
-				err = errors.Join(err, fmt.Errorf(stat.Error))
+				err = errors.Join(err, fmt.Errorf("%w", stat.Error))
```
Comment (josvazg) [inline pkg/controller/atlasproject/custom_roles.go:203]:
```suggestion
				err = errors.Join(err, fmt.Errorf("%s", stat.Error))
```

File: pkg/controller/atlasproject/custom_roles.go [hunk 5]
```diff
@@ -210,7 +210,7 @@ func syncCustomRolesStatus(ctx *workflow.Context, desiredCustomRoles []customrol
 			statuses = append(statuses, stat)
 
 			if stat.Status == status.CustomRoleStatusFailed {
-				err = errors.Join(err, fmt.Errorf(stat.Error))
+				err = errors.Join(err, fmt.Errorf("%w", stat.Error))
```
Comment (josvazg) [inline pkg/controller/atlasproject/custom_roles.go:213]: stat.Error is a `string`

```suggestion
				err = errors.Join(err, fmt.Errorf("%s", stat.Error))
```

File: pkg/controller/atlasproject/custom_roles.go [hunk 6]
```diff
@@ -220,7 +220,7 @@ func syncCustomRolesStatus(ctx *workflow.Context, desiredCustomRoles []customrol
...
```
Comment (josvazg) [inline pkg/controller/atlasproject/custom_roles.go:223]: stat.Error is a `string`

```suggestion
				err = errors.Join(err, fmt.Errorf("%s", stat.Error))
```

**Review by s-urbaniak — APPROVED** [review]

**Review by helderjs — APPROVED** [review]

**Pull request merged**